File: go/bob/bob.go
/*
Package bob implements a very, very simple chatbot.
*/
package bob

import (
	"regexp"
	"strings"
)

// Hey responds to remarks in a very simple way.
func Hey(remark string) string {
	switch {
	case isQuestion(&remark) && isUpper(&remark):
		return "Calm down, I know what I'm doing!"
	case isUpper(&remark):
		return "Whoa, chill out!"
	case isQuestion(&remark):
		return "Sure."
	case isSilence(&remark):
		return "Fine. Be that way!"
	default:
		return "Whatever."
	}
}

func isUpper(s *string) bool {
	words, _ := regexp.MatchString("[a-zA-z]", *s)
	return strings.ToUpper(*s) == *s && words
}

func isQuestion(s *string) bool {
	question, _ := regexp.MatchString("\\? *$", *s)
	return question
}

func isSilence(s *string) bool {
	silence, _ := regexp.MatchString("^[ \\t\\n\\r]*$", *s)
	return silence
}

File: elixir/bob/bob.exs
defmodule Bob do
  def hey(input) do
    cond do
      # Saying nothing
      nothing?(input) ->
        "Fine. Be that way!"

      # Yelling a question
      yelling?(input) && asking?(input) ->
        "Calm down, I know what I'm doing!"

      # Yelling
      yelling?(input) ->
        "Whoa, chill out!"

      # Asking
      asking?(input) ->
        "Sure."

      # Everything else
      true ->
        "Whatever."
    end
  end

  defp asking?(input), do: String.ends_with?(input, "?")
  defp nothing?(input), do: String.split(input) == []
  defp yelling?(input), do: input == String.upcase(input) && has_letters?(input)
  defp has_letters?(input), do: Regex.match?(~r(\pL), input)
end
